### PR TITLE
Feature/authz tp

### PR DESCRIPTION
**Pull request by ryanbtp**
An OPA policy and key registration transaction processor, tested at the
    protobuf level. Similar in design to the chronicle TP, but should have
    no dependencies on the rest of the Chronicle project
    
    Negative / property testing required for cases that cannot be created
    with the client message builders.
    
    Don't review opactl, that's very much WIP
    
    References:
    
    CHRON-201 CHRON-183

### CODE REVIEW

File: crates/opa-tp-protocol/src/sawtooth.rs [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+    #[test]
+    fn sawtooth_batch_roundtrip() {
+        let signing_key = key_from_seed(0);
+
+        let submission = SubmissionBuilder::bootstrap_root(signing_key.verifying_key()).build(1);
+        let mut builder = MessageBuilder::new(signing_key, "external_id", "version");
+
+        let input_addresses = vec!["inone".to_owned(), "intwo".to_owned()];
+        let output_addresses = vec!["outtwo".to_owned(), "outtwo".to_owned()];
```
Comment (suchapalaver) [inline crates/opa-tp-protocol/src/sawtooth.rs:142]:
```suggestion
        let output_addresses = vec!["outone".to_owned(), "outtwo".to_owned()];
```

File: crates/api/Cargo.toml [hunk 2]
```diff
@@ -17,6 +17,7 @@ async-graphql-poem = "4.0.9"
 async-stream = "0.3.3"
 async-trait = "0.1.53"
 chrono = "0.4.19"
+
```
Comment (mtbc) [inline crates/api/Cargo.toml:20]: unintended?

File: crates/opa-tp/src/tp.rs [hunk 3]
```diff
@@ -0,0 +1,974 @@
...
+        OpaTransactionHandler {
+            family_name: FAMILY.to_owned(),
+            family_versions: vec![VERSION.to_owned()],
+            namespaces: vec![PREFIX.to_string()],
+        }
+    }
+}
+
+// Verifies the submission.
+// Keys == Non indicates that the opa tp is not bootstrapped, so the bootstrap
```
Comment (mtbc) [inline crates/opa-tp/src/tp.rs:57]:
```suggestion
// Keys == None indicates that the opa tp is not bootstrapped, so the bootstrap
```

File: crates/opactl/src/cli.rs [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+                .env("NEW_ROOT_KEY")
+                .num_args(1)
+                .value_hint(ValueHint::FilePath)
+                .help("The new key to register as the root key"),
+        )
+}
+
+pub fn register_key() -> Command {
+    Command::new("register-key")
+        .about("Register a new non root key with the OPA transaction processor")
```
Comment (mtbc) [inline crates/opactl/src/cli.rs:54]:
```suggestion
        .about("Register a new non-root key with the OPA transaction processor")
```

**Review by suchapalaver — APPROVED** [review]

File: crates/opa-tp-protocol/src/sawtooth.rs [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+#[derive(Debug, Clone)]
+pub struct MessageBuilder {
+    signer: SigningKey,
+    family_name: String,
+    family_version: String,
+    rng: StdRng,
+}
+
+impl MessageBuilder {
+    #[allow(dead_code)]
```
Comment (mtbc) [inline crates/opa-tp-protocol/src/sawtooth.rs:26]: It's curious that you need this.

File: crates/opa-tp/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+            _ => ConsoleLogging::Off,
+        },
+    );
+
+    let handler = OpaTransactionHandler::new();
+    let mut processor = TransactionProcessor::new({
+        if let Some(connect) = matches.get_one::<String>("connect") {
+            connect
+        } else {
+            "tcp://127.0.0.1:4004"
```
Comment (mtbc) [inline crates/opa-tp/src/main.rs:69]: maybe instead set default on `Arg`

File: crates/opactl/src/cli.rs [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+                .value_parser(NonEmptyStringValueParser::new())
+                .default_value("default")
+                .help("The id of the policy, if not specified then the default policy is returned"),
+        )
+}
+
+pub fn cli() -> Command {
+    Command::new("opactl")
+        .version(env!("CARGO_PKG_VERSION"))
+        .author("BTP.works TODO WHAT IS THIS NOW")
```
Comment (mtbc) [inline crates/opactl/src/cli.rs:170]: I guess we're getting back to this one. :smile:

**Comment by mtbc** [comment]
:+1: for the diagrams! The architectural overview's helpful.

**Pull request merged**